Packages/com.unity.inputsystem/InputSystem/InputManager.cs

```diff
@@ -3739,7 +3773,8 @@ internal unsafe bool UpdateState(InputDevice device, InputEvent* eventPtr, Input
                     deviceStateSize);
             }
 
-            if (InputSettings.readValueCachingFeatureEnabled)
+            // Keyboard has enough ButtonControls that it's faster to find out which have actually changed here.
+            if (InputSettings.readValueCachingFeatureEnabled || m_Devices[deviceIndex] is Keyboard)
```

Comment:
I wonder if there is also a synergy with knowing the button count also here? It sounds like the benefit here is actually more related to the number of buttons on the device rather than it being a Keyboard? Hence maybe this should be a threshold based off number of buttons? Its fine as is but something like that would make it more future proof?

Reply:
Updated - with some testing with the new changes, it seems that having about 45 buttons that have pressed/releasedThisFrame usage is where it becomes more efficient to take the cache value, so we now swap over at that point.